src/middlewared/middlewared/plugins/kmip/sync.py

```diff
@@ -0,0 +1,363 @@
+                else:
+                    continue
+            except Exception:
+                failed.append(ds['name'])
+            else:
+                update_data = {'encryption_key': self.middleware.call_sync('pwenc.encrypt', key), 'kmip_uid': None}
+                self.middleware.call_sync('datastore.update', 'storage.encrypteddataset', ds['id'], update_data)
+                self.zfs_keys.pop(ds['name'], None)
+                if connection_successful:
+                    with self._connection(self.connection_config()) as conn:
+                        self._revoke_and_destroy_key(ds['kmip_uid'], conn)
+        self.zfs_keys = {k: v for k, v in self.zfs_keys.items() if k in existing_datasets}
+        return failed
+
+    @private
+    @job(lock=lambda args: f'kmip_sync_zfs_keys_{args}')
+    def sync_zfs_keys(self, job, ids=None):
+        if not self.middleware.call_sync('kmip.zfs_keys_pending_sync'):
+            return
+        config = self.middleware.call_sync('kmip.config')
+        conn_successful = self.test_connection(raise_alert=True)
+        if config['enabled'] and config['manage_zfs_keys']:
+            if conn_successful:
+                failed = self.push_zfs_keys(ids)
+            else:
+                return
+        else:
+            failed = self.pull_zfs_keys()
+        if failed:
+            self.middleware.call_sync(
+                'alert.oneshot_create', 'KMIPZFSDatasetsSyncFailure', {'datasets': ','.join(failed)}
+            )
+
+    @periodic(interval=86400)
+    async def sync_keys(self):
+        if not await self.middleware.call('kmip.kmip_sync_pending'):
+            return
+        await self.middleware.call('kmip.sync_zfs_keys')
+        await self.middleware.call('kmip.sync_sed_keys')
+
+    @private
+    def push_sed_keys(self, ids=None):
+        adv_config = self.middleware.call_sync('system.advanced.config')
+        failed = []
+        with self._connection(self.connection_config()) as conn:
+            for disk in self.middleware.call_sync('disk.query', [['id', 'in' if ids else 'nin', ids or []]]):
+                if not disk['passwd'] and disk['kmip_uid']:
+                    try:
+                        key = self._retrieve_secret_data(disk['kmip_uid'], conn)
+                    except Exception as e:
+                        self.middleware.logger.debug(f'Failed to retrieve key for {disk["identifier"]}: {e}')
+                    else:
+                        self.disks_keys[disk['identifier']] = key
+                    continue
+                elif not disk['passwd']:
+                    continue
+
+                self.disks_keys[disk['identifier']] = disk['passwd']
+                destroy_successful = False
+                if disk['kmip_uid']:
+                    # This needs to be revoked and destroyed
+                    destroy_successful = self._revoke_and_destroy_key(disk['kmip_uid'], conn)
```

Comment:
Isnt there anything we can log how/why revoke and destroy failed? I didnt see it in the call chain, maybe I am missing something